### PR TITLE
Beam/sdks/io/gcp/java/healcare/hl7v2 io read

### PR DESCRIPTION
**Pull request by kamrankoupayi**
Adding HL7v2ReadParameter and HL7v2ReadResponse which allow to sent metadata while reading from HL7v2Store, and a function to read from PCollection of HL7v2ReadParameter and return a result which contains PCollection messages of HLReadResponse and PCollection failed results of HealthCareIOError<HL7v2Parameter>.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kamrankoupayi** [comment]
R: @pabloem

**Comment by kamrankoupayi** [comment]
R: @ahmedabu98

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2ReadResponse.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+package org.apache.beam.sdk.io.gcp.healthcare;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * HL7v2ReadResponse represents the response format for a HL7v2 read request, used as a parameter
+ * for {@link HL7v2IO.Read}.
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2ReadResponse.java:27]:
```suggestion
 * HL7v2ReadResponse represents the response format for a HL7v2 read request, used as the output
 * type of {@link HL7v2IO.HL7v2Read}.
```

**Review by ahmedabu98 — APPROVED** [review]
Just left a couple suggestions about documentation, feel free to ignore if they don't make sense. Other than that, this PR looks great!

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IOTest.java [hunk 2]
```diff
@@ -59,6 +61,30 @@ public void test_HL7v2IO_failedReads() {
     pipeline.run();
   }
 
+  @Test
+  public void test_HL7v2IO_failedReadsByParameter() {
```
Comment (Abacn) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IOTest.java:65]:
```suggestion
  public void testHL7v2IOFailedReadsByParameter() {
```

We use smallCamelNamingStyle in Java

This leads to me to wonder if our checkStyle for test code no longer working.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IO.java [hunk 3]
```diff
@@ -260,15 +266,12 @@ public static class Read extends PTransform<PCollection<String>, Read.Result> {
 
     public Read() {}
 
-    public static class Result implements POutput, PInput {
-
-      private PCollection<HL7v2Message> messages;
-
-      private PCollection<HealthcareIOError<String>> failedReads;
-      PCollectionTuple pct;
+    public static class Result extends HL7v2ReadResult<String, HL7v2Message> {
```
Comment (Abacn) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IO.java:269]: ideally each public class (that would be rendered to javadoc) should have docstring. Would you mind add a brief docstring also here?

**Review by Abacn — COMMENTED** [review]
Thanks, looks pretty good. Got a couple of minor comment.

Also, the added integration test has a warning: 

https://ci-beam.apache.org/job/beam_PreCommit_Java_GCP_IO_Direct_Commit/1564/testReport/org.apache.beam.sdk.io.gcp.healthcare/HL7v2IOReadWriteIT/testHL7v2IOGetAllE2E/

```
WARNING: Can't verify serialized elements of type HL7v2ReadParameter have well defined equals method. This may produce incorrect results on some PipelineRunner implementations
```

Is this something need attention?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IOTest.java (hunk 2 above)
Comment (kamrankoupayi) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IOTest.java]: So, I've change the names of the tests to be smallCamelNamingStyle in here.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IO.java (hunk 3 above)
Comment (kamrankoupayi) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IO.java:273]: Thanks for mentioning, added  brief docstring.

**Comment by kamrankoupayi** [comment]
> Thanks, looks pretty good. Got a couple of minor comment.
> 
> Also, the added integration test has a warning:
> 
> https://ci-beam.apache.org/job/beam_PreCommit_Java_GCP_IO_Direct_Commit/1564/testReport/org.apache.beam.sdk.io.gcp.healthcare/HL7v2IOReadWriteIT/testHL7v2IOGetAllE2E/
> 
> ```
> WARNING: Can't verify serialized elements of type HL7v2ReadParameter have well defined equals method. This may produce incorrect results on some PipelineRunner implementations
> ```
> 
> Is this something need attention?

Thank you for review and your comments.
And yes that's important so I added the equal function for HL7v2ReadParameter class to address this warning. Thanks for catching.

**Review by Abacn — APPROVED** [review]
Thanks, this LGTM

**Pull request merged**